runtime/parachains/src/paras.rs

```diff
@@ -524,9 +665,174 @@ pub mod pallet {
 			Self::deposit_event(Event::ActionQueued(para, next_session));
 			Ok(())
 		}
+
+		/// Includes a statement for a PVF pre-checking vote. Potentially, finalizes the vote and
+		/// enacts the results if that was the last vote before achieving the supermajority.
+		#[pallet::weight(0)]
+		pub fn include_pvf_check_statement(
+			origin: OriginFor<T>,
+			stmt: PvfCheckStatement,
+			signature: ValidatorSignature,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+			let validators = shared::Pallet::<T>::active_validator_keys();
+			let current_session = shared::Pallet::<T>::session_index();
+			if stmt.session_index < current_session {
+				return Err(Error::<T>::PvfCheckStatementStale.into())
+			} else if stmt.session_index > current_session {
+				return Err(Error::<T>::PvfCheckStatementFuture.into())
+			}
+			let validator_index = stmt.validator_index.0 as usize;
+			let validator_public = validators
+				.get(validator_index)
+				.ok_or(Error::<T>::PvfCheckValidatorIndexOutOfBounds)?;
+
+			let signing_payload = stmt.signing_payload();
+			ensure!(
+				signature.verify(&signing_payload[..], &validator_public),
+				Error::<T>::PvfCheckInvalidSignature,
+			);
+
+			let mut active_vote = PvfActiveVoteMap::<T>::get(&stmt.subject)
+				.ok_or(Error::<T>::PvfCheckSubjectInvalid)?;
+
+			ensure!(
```

Comment:
:+1: a comment about the importance of this one preventing votes in both directions, would be nice.

Reply:
Well, I tried to come up with a comment and couldn't. That does not seem to be that important, since everyone sees each other's vote on-chain. Perhaps it just allows for wasting resources.